### PR TITLE
[Merged by Bors] - Remove sync await points

### PR DESCRIPTION
**Pull request by divagant-martian**
## Issue Addressed

Removes the await points in sync waiting for a processor response for rpc block processing. Built on top of #3029 
This also handles a couple of bugs in the previous code and adds a relatively comprehensive test suite.

### CODE REVIEW

**Comment by divagant-martian** [comment]
@AgeManning this is running in Toronto to dissuade some doubts I have, but it's otherwise ready. Multiple peers per requests is in the next pr

**Comment by AgeManning** [comment]
Could we merge latest unstable to simplify the diff?

File: beacon_node/network/src/beacon_processor/worker/sync_methods.rs [hunk 1]
```diff
@@ -39,76 +39,101 @@ impl<T: BeaconChainTypes> Worker<T> {
...
-        // Checks if the block is already being imported through another source
-        if let Some(handle) = duplicate_cache.check_and_insert(block_root) {
-            let slot = block.slot();
-            let block_result = self.chain.process_block(block);
+        // Check if the block is already being imported through another source
+        let root = block.canonical_root();
+        let handle = match duplicate_cache.check_and_insert(root) {
+            Some(handle) => handle,
+            None => {
+                return;
```
Comment (AgeManning) [inline beacon_node/network/src/beacon_processor/worker/sync_methods.rs:55]: I'm not sure if these cases are handled, but I would imagine that when we are requesting blocks to be processed, there is some state or mapping keeping track as to whether the blocks get processed successfully or fail, in parent lookup or single block lookup. Potentially, for this function its not the case, but if it is, I think we need to always emit a response indicating the block failed processing, to handle it by whoever requested it.
Here we just return.

File: beacon_node/network/src/beacon_processor/worker/sync_methods.rs [hunk 2]
```diff
@@ -39,76 +39,101 @@ impl<T: BeaconChainTypes> Worker<T> {
...
-            // Here, we assume that the block will eventually be imported and
-            // send a `BlockIsAlreadyKnown` message to sync.
-            if result_tx
-                .send(Err(BlockError::BlockIsAlreadyKnown))
-                .is_err()
-            {
-                crit!(self.log, "Failed return sync block result");
+            (BlockProcessType::SingleBlock { .. }, Err(e)) => match e {
+                BlockError::BlockIsAlreadyKnown => {
+                    // No error here
```
Comment (AgeManning) [inline beacon_node/network/src/beacon_processor/worker/sync_methods.rs]: here we also just return

File: beacon_node/network/src/beacon_processor/worker/sync_methods.rs [hunk 3]
```diff
@@ -39,76 +39,101 @@ impl<T: BeaconChainTypes> Worker<T> {
...
+                }
+                BlockError::BeaconChainError(e) => {
+                    // Internal error
+                    error!(self.log, "Beacon chain error processing single block"; "block_root" => %root, "error" => ?e);
```
Comment (AgeManning) [inline beacon_node/network/src/beacon_processor/worker/sync_methods.rs]: also here

File: beacon_node/network/src/beacon_processor/worker/sync_methods.rs [hunk 4]
```diff
@@ -39,76 +39,101 @@ impl<T: BeaconChainTypes> Worker<T> {
     pub fn process_rpc_block(
```
Comment (AgeManning) [inline beacon_node/network/src/beacon_processor/worker/sync_methods.rs:38]: Probably should update the comment for this function. Previously we returned a result always to the `result_tx` channel, but now we only sometimes send a result to sync

**Review by AgeManning — APPROVED** [review]
I'm keen to get this is. I've not extensively reviewed this, because the changes are substantial, but if there is anything off, we should notice it in our testing.

**Comment by divagant-martian** [comment]
Let's wait a bit more on this

**Comment by divagant-martian** [comment]
Ready again thanks to @pawanjay176

**Comment by divagant-martian** [comment]
@michaelsproul why is this blocked?

**Comment by michaelsproul** [comment]
> why is this blocked?

was blocked on the v2.1.4 release but should be good to go now I think!

**Comment by michaelsproul** [comment]
looks like the Python versioning issue that @realbigsean recently fixed in #3056. not sure why it would be inconsistent, let's try again

bors retry

**Comment by michaelsproul** [comment]
Now that https://github.com/sigp/lighthouse/pull/3105 is merged let's try again

bors retry

**Comment by michaelsproul** [comment]
:pray:

bors +

**Pull request closed without merging**